README
NAME
    pod2xml - Script to convert PYX to XML.

SYNOPSIS
     pyx2xml [-e in_enc] [-h] [-i] [-s no_simple] [--version] [filename] [-]

DESCRIPTION
     pyx2xml script uses PYX::SGML::Tags class to create XML output of PYX.

ARGUMENTS
    *       "-e in_enc"

             Input encoding.
             Default value is 'utf-8'.

    *       "-h"

             Print help.

    *       "-i"

             Indent output.

    *       "-s no_simple"

             List of element, which cannot be a simple like <element/>.
             Separator is comma.

    *       "--version"

             Print version of script.

    *       "filename"

             Input file in PYX format.

    *       "-"

             Stdin with PYX input.

EXAMPLE1
     pyx2xml

     # Output:
     # Usage: pyx2xml [-e in_enc] [-h] [-i] [--version] [filename] [-]
     #         -e in_enc       Input encoding (default value is utf-8).
     #         -h              Print help.
     #         -i              Indent output.
     #         --version       Print version.
     #         [filename]      Process on filename
     #         [-]             Process on stdin

EXAMPLE2
     cat <<END | pyx2xml -
     (element
     -foo
     )element
     END

     # Output:
     # <element>foo</element>

EXAMPLE3
     cat <<END | pyx2xml -s foo,baz -
     (foo
     )foo
     (bar
     )bar
     (baz
     )baz
     END

     # Output:
     # <foo></foo><bar /><baz></baz>

REPOSITORY
    <https://github.com/michal-josef-spacek/App-PYX2XML>

AUTHOR
    Michal Josef Špaček <mailto:skim@cpan.org>

    <http://skim.cz>

LICENSE AND COPYRIGHT
    © 2015-2022 Michal Josef Špaček

    BSD 2-Clause License

VERSION
    0.07



Changes
0.07
 - Fix Makefile.PL for new version of Perl and no default »use lib '.'«.
 - Rename example script to better name.
 - Update Module::Install to 1.21 version.
 - Update copyright years.

0.06 2021-02-05T12:20:42+01:00
 - Update README.
 - Update examples/ dir.

0.05 2021-02-05T12:19:16+01:00
 - Add -s option to script.
 - Add tests for SGML and not ended element.
 - Fix bugtracker.
 - Fix exit code.
 - Update copyright years.

0.04 2020-04-26T22:52:34+02:00
 - Fix test dependencies.
 - Remove obsolete ERRORS section in doc.
 - Rewrite doc METHODS section to =head2.

0.03 2020-04-26T19:51:53+02:00
 - Add test for constructor.
 - Add test for run() method.
 - Fix usage of -e option.
 - Fix to 80 characters on line.
 - Improve LICENSE AND COPYRIGHT section in doc.
 - Improve SYNOPSIS section in doc.
 - Move argument parsing from constructor to run section.
 - Sort attributes.
 - Update Module::Install to 1.19 version.
 - Update author github username.
 - Update author name.
 - Update copyright years.

0.02 2016-02-20T23:43:25+01:00
 - Add -e option for input encoding.

0.01 2016-02-17T11:13:38+01:00
 - First version.


PYX2XML.pm
package App::PYX2XML;

use strict;
use warnings;

use Getopt::Std;
use PYX::SGML::Tags;
use Tags::Output::Indent;
use Tags::Output::Raw;

our $VERSION = 0.07;

# Constructor.
sub new {
	my ($class, @params) = @_;

	# Create object.
	my $self = bless {}, $class;

	# Object.
	return $self;
}

# Run script.
sub run {
	my $self = shift;

	# Process arguments.
	$self->{'_opts'} = {
		'e' => 'utf-8',
		'h' => 0,
		'i' => 0,
		's' => '',
	};
	if (! getopts('e:his:', $self->{'_opts'}) || $self->{'_opts'}->{'h'}
		|| @ARGV < 1) {

		print STDERR "Usage: $0 [-e in_enc] [-h] [-i] [-s no_simple] [--version] ".
			"[filename] [-]\n";
		print STDERR "\t-e in_enc\tInput encoding (default value ".
			"is utf-8)\n";
		print STDERR "\t-h\t\tPrint help.\n";
		print STDERR "\t-i\t\tIndent output.\n";
		print STDERR "\t-s no_simple\tList of element, which cannot be a simple".
			" like <element/>. Separator is comma.\n";
		print STDERR "\t--version\tPrint version.\n";
		print STDERR "\t[filename]\tProcess on filename\n";
		print STDERR "\t[-]\t\tProcess on stdin\n";
		return 1;
	}
	$self->{'_filename_or_stdin'} = $ARGV[0];

	# No simple elements.
	my @no_simple = split m/,/ms, $self->{'_opts'}->{'s'};

	# Tags object.
	my $tags;
	my %params = (
		'no_simple' => \@no_simple,
		'output_handler' => \*STDOUT,
		'xml' => 1,
	);
	if ($self->{'_opts'}->{'i'}) {
		$tags = Tags::Output::Indent->new(%params);
	} else {
		$tags = Tags::Output::Raw->new(%params);
	}

	# PYX::SGML::Tags object.
	my $writer = PYX::SGML::Tags->new(
		'input_encoding' => $self->{'_opts'}->{'e'},
		'tags' => $tags,
	);

	# Parse from stdin.
	if ($self->{'_filename_or_stdin'} eq '-') {
		$writer->parse_handler(\*STDIN);

	# Parse file.
	} else {
		$writer->parse_file($self->{'_filename_or_stdin'});
	}
	print "\n";

	return 0;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

App::PYX2XML - Perl class for pyx2xml application.

=head1 SYNOPSIS

 use App::PYX2XML;

 my $obj = App::PYX2XML->new;
 $obj->run;

=head1 METHODS

=head2 C<new>

 my $obj = App::PYX2XML->new;

Constructor.

Returns instance of object.

=head2 C<run>

 $obj->run;

Run.

Returns undef.

=head1 EXAMPLE

=for comment filename=print_help.pl

 use strict;
 use warnings;

 use App::PYX2XML;

 # Run.
 exit App::PYX2XML->new->run;

 # Output:
 # Usage: ./examples/ex1.pl [-e in_enc] [-h] [-i] [-s no_simple] [--version] [filename] [-]
 #         -e in_enc       Input encoding (default value is utf-8).
 #         -h              Print help.
 #         -i              Indent output.
 #         -s no_simple    List of element, which cannot be a simple like <element/>. Separator is comma.
 #         --version       Print version.
 #         [filename]      Process on filename
 #         [-]             Process on stdin

=head1 DEPENDENCIES

L<Getopt::Std>,
L<PYX::SGML::Tags>,
L<Tags::Output::Indent>,
L<Tags::Output::Raw>.

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/App-PYX2XML>.

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© 2015-2022 Michal Josef Špaček

BSD 2-Clause License

=head1 VERSION

0.07

=cut


Makefile.PL
use lib '.';
use strict;
use warnings;

use inc::Module::Install;

# Definition.
abstract 'Perl class for pyx2xml application.';
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'English' => 0;
author_requires 'File::Object' => 0.08;
author_requires 'Test::More' => 0;
author_requires 'Test::NoWarnings' => 0;
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
install_script 'bin/pyx2xml';
license 'bsd';
name 'App-PYX2XML';
readme_from 'bin/pyx2xml';
recursive_author_tests('xt');
requires 'Getopt::Std' => 0;
requires 'PYX::SGML::Tags' => 0.02;
requires 'Tags::Output::Indent' => 0.04;
requires 'Tags::Output::Raw' => 0.04;
requires 'perl' => '5.6.2';
resources 'bugtracker' => 'https://github.com/michal-josef-spacek/App-PYX2XML/issues';
resources 'homepage' => 'https://github.com/michal-josef-spacek/App-PYX2XML';
resources 'repository' => 'git://github.com/michal-josef-spacek/App-PYX2XML';
test_requires 'English' => 0;
test_requires 'Error::Pure::Utils' => 0;
test_requires 'File::Object' => 0;
test_requires 'Unicode::UTF8' => 0;
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
test_requires 'Test::Output' => 0;
tests_recursive;
version '0.07';

# Run.
WriteAll();
